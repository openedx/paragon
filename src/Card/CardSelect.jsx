import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Card from 'react-bootstrap/Card';
import Form from '../Form';

const CardSelect = ({
  className,
  cardsData,
  onChange,
  selectedCardName,
}) => {
  const [currentRecord, setCurrentRecord] = useState(null);
  useEffect(() => {
    setCurrentRecord(selectedCardName);
  }, []);

  return (
    <>
      <Form.Group>
        <Form.RadioSet
          name="radio-select-card"
          onChange={(e) => {
            setCurrentRecord(e.target.value);
            onChange(e.target.value);
          }}
          defaultValue={selectedCardName}
          isInline
          className="row"
        >
          {cardsData?.map(data => (
            <div
              key={data.name}
              className={className}
              role="group"
              aria-label={data.title}
            >
              <Card className={currentRecord === data.name ? 'border border-dark' : null}>
                <Card.Img variant="top" src={data.img} />
                <Card.Body>
                  <Card.Title as="h5" className="card-title mb-1">
                    <label htmlFor={data.name}>{data.title}</label>
                    <div className="float-right"><Form.Radio id={data.name} value={data.name} /></div>
                  </Card.Title>
                  <Card.Text>
                    {data.textElements?.map(ele => (
                      <p className={ele.className}>{ele.text}</p>
                    ))}
                  </Card.Text>
                </Card.Body>
                <Card.Footer>
                  <small className="text-muted">{data.footer}</small>
                </Card.Footer>
              </Card>
            </div>
          ))}
        </Form.RadioSet>
      </Form.Group>
    </>
  );
};

CardSelect.defaultProps = {
  className: 'mb-3 col-md-3 offset-md-1',
  onChange: () => {},
};

CardSelect.propTypes = {
  /** The class name for the CardSelect component */
  className: PropTypes.string,
  /** The Card components data to organize into a radio selectable form */
  cardsData: PropTypes.node.isRequired,
  /** specifies the callback for the onChange event. The default value is a no-op function. */
  onChange: PropTypes.func,
  /** The selected card */
  selectedCardName: PropTypes.string.isRequired,
};

export default CardSelect;
