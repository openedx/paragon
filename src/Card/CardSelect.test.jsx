import React from 'react';
import renderer from 'react-test-renderer';
import CardSelect from './CardSelect';

describe('<CardSelect />', () => {
  describe('First card selected', () => {
    it('First card selected', () => {
      const tree = renderer.create((
        <CardSelect
          cardsData={[
            { name: 'card1', title: 'Lorem Ipsum 1' },
            { name: 'card2', title: 'Lorem Ipsum 2', img: 'https://source.unsplash.com/360x200/?nature' },
            { name: 'card3', title: 'Lorem Ipsum 3', img: 'https://source.unsplash.com/360x200' }]}
          selectedCardName="card1"
        />
      )).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Second card selected', () => {
    it('Second card selected', () => {
      const tree = renderer.create((
        <CardSelect
          cardsData={[
            { name: 'card1', title: 'Lorem Ipsum 1' },
            { name: 'card2', title: 'Lorem Ipsum 2', img: 'https://source.unsplash.com/360x200/?nature' },
            { name: 'card3', title: 'Lorem Ipsum 3', img: 'https://source.unsplash.com/360x200' }]}
          selectedCardName="card2"
        />
      )).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
