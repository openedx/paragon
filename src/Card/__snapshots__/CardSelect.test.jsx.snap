// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardSelect /> First card selected First card selected 1`] = `
<div
  className="pgn__form-group"
>
  <div
    className="row pgn__form-control-set-inline"
    id="form-field1"
    role="radiogroup"
  >
    <div
      aria-label="Lorem Ipsum 1"
      className="mb-3 col-md-3 offset-md-1"
      role="group"
    >
      <div
        className="card"
      >
        <img
          className="card-img-top"
        />
        <div
          className="card-body"
        >
          <h5
            className="card-title mb-1 card-title"
          >
            <label
              htmlFor="card1"
            >
              Lorem Ipsum 1
            </label>
            <div
              className="float-right"
            >
              <div
                className="pgn__form-radio"
              >
                <input
                  className="pgn__form-radio-input"
                  defaultChecked={true}
                  id="card1"
                  name="radio-select-card"
                  onChange={[Function]}
                  type="radio"
                  value="card1"
                />
                <div>
                  <label
                    className="pgn__form-label"
                    htmlFor="card1"
                  />
                </div>
              </div>
            </div>
          </h5>
          <p
            className="card-text"
          />
        </div>
        <div
          className="card-footer"
        >
          <small
            className="text-muted"
          />
        </div>
      </div>
    </div>
    <div
      aria-label="Lorem Ipsum 2"
      className="mb-3 col-md-3 offset-md-1"
      role="group"
    >
      <div
        className="card"
      >
        <img
          className="card-img-top"
          src="https://source.unsplash.com/360x200/?nature"
        />
        <div
          className="card-body"
        >
          <h5
            className="card-title mb-1 card-title"
          >
            <label
              htmlFor="card2"
            >
              Lorem Ipsum 2
            </label>
            <div
              className="float-right"
            >
              <div
                className="pgn__form-radio"
              >
                <input
                  className="pgn__form-radio-input"
                  defaultChecked={false}
                  id="card2"
                  name="radio-select-card"
                  onChange={[Function]}
                  type="radio"
                  value="card2"
                />
                <div>
                  <label
                    className="pgn__form-label"
                    htmlFor="card2"
                  />
                </div>
              </div>
            </div>
          </h5>
          <p
            className="card-text"
          />
        </div>
        <div
          className="card-footer"
        >
          <small
            className="text-muted"
          />
        </div>
      </div>
    </div>
    <div
      aria-label="Lorem Ipsum 3"
      className="mb-3 col-md-3 offset-md-1"
      role="group"
    >
      <div
        className="card"
      >
        <img
          className="card-img-top"
          src="https://source.unsplash.com/360x200"
        />
        <div
          className="card-body"
        >
          <h5
            className="card-title mb-1 card-title"
          >
            <label
              htmlFor="card3"
            >
              Lorem Ipsum 3
            </label>
            <div
              className="float-right"
            >
              <div
                className="pgn__form-radio"
              >
                <input
                  className="pgn__form-radio-input"
                  defaultChecked={false}
                  id="card3"
                  name="radio-select-card"
                  onChange={[Function]}
                  type="radio"
                  value="card3"
                />
                <div>
                  <label
                    className="pgn__form-label"
                    htmlFor="card3"
                  />
                </div>
              </div>
            </div>
          </h5>
          <p
            className="card-text"
          />
        </div>
        <div
          className="card-footer"
        >
          <small
            className="text-muted"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CardSelect /> Second card selected Second card selected 1`] = `
<div
  className="pgn__form-group"
>
  <div
    className="row pgn__form-control-set-inline"
    id="form-field5"
    role="radiogroup"
  >
    <div
      aria-label="Lorem Ipsum 1"
      className="mb-3 col-md-3 offset-md-1"
      role="group"
    >
      <div
        className="card"
      >
        <img
          className="card-img-top"
        />
        <div
          className="card-body"
        >
          <h5
            className="card-title mb-1 card-title"
          >
            <label
              htmlFor="card1"
            >
              Lorem Ipsum 1
            </label>
            <div
              className="float-right"
            >
              <div
                className="pgn__form-radio"
              >
                <input
                  className="pgn__form-radio-input"
                  defaultChecked={false}
                  id="card1"
                  name="radio-select-card"
                  onChange={[Function]}
                  type="radio"
                  value="card1"
                />
                <div>
                  <label
                    className="pgn__form-label"
                    htmlFor="card1"
                  />
                </div>
              </div>
            </div>
          </h5>
          <p
            className="card-text"
          />
        </div>
        <div
          className="card-footer"
        >
          <small
            className="text-muted"
          />
        </div>
      </div>
    </div>
    <div
      aria-label="Lorem Ipsum 2"
      className="mb-3 col-md-3 offset-md-1"
      role="group"
    >
      <div
        className="border border-dark card"
      >
        <img
          className="card-img-top"
          src="https://source.unsplash.com/360x200/?nature"
        />
        <div
          className="card-body"
        >
          <h5
            className="card-title mb-1 card-title"
          >
            <label
              htmlFor="card2"
            >
              Lorem Ipsum 2
            </label>
            <div
              className="float-right"
            >
              <div
                className="pgn__form-radio"
              >
                <input
                  className="pgn__form-radio-input"
                  defaultChecked={true}
                  id="card2"
                  name="radio-select-card"
                  onChange={[Function]}
                  type="radio"
                  value="card2"
                />
                <div>
                  <label
                    className="pgn__form-label"
                    htmlFor="card2"
                  />
                </div>
              </div>
            </div>
          </h5>
          <p
            className="card-text"
          />
        </div>
        <div
          className="card-footer"
        >
          <small
            className="text-muted"
          />
        </div>
      </div>
    </div>
    <div
      aria-label="Lorem Ipsum 3"
      className="mb-3 col-md-3 offset-md-1"
      role="group"
    >
      <div
        className="card"
      >
        <img
          className="card-img-top"
          src="https://source.unsplash.com/360x200"
        />
        <div
          className="card-body"
        >
          <h5
            className="card-title mb-1 card-title"
          >
            <label
              htmlFor="card3"
            >
              Lorem Ipsum 3
            </label>
            <div
              className="float-right"
            >
              <div
                className="pgn__form-radio"
              >
                <input
                  className="pgn__form-radio-input"
                  defaultChecked={false}
                  id="card3"
                  name="radio-select-card"
                  onChange={[Function]}
                  type="radio"
                  value="card3"
                />
                <div>
                  <label
                    className="pgn__form-label"
                    htmlFor="card3"
                  />
                </div>
              </div>
            </div>
          </h5>
          <p
            className="card-text"
          />
        </div>
        <div
          className="card-footer"
        >
          <small
            className="text-muted"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
